docs/langref/relay_expr.rst

```diff
@@ -0,0 +1,402 @@
+
+Variables
+=========
+
+Relay allows for local and global variables. Our design is based on
+that of LLVM, which differentiates between identifier types; a writer of
+optimizations can thus determine a lot of information about what a
+variable references simply by knowing the kind of identifier.
+
+Global variables are written with `@`, local variables are written
+with `%`, and variables written without a sigil correspond to operator
+names.
+
+The distinction between global and local identifiers
+makes certain kinds of transformation easier. For example,
+inlining a global definition requires no analysis: simply inlining
+the definitions suffices.
+
+Global Variable
+~~~~~~~~~~~~~~~~~~
+
+Global identifiers are prefixed by the `@` sigil, such as "`@global`".
+A global identifier always references a globally visibly definition contained in the environment. The names of global identifiers must be unique.
+
+See :py:class:`~tvm.relay.expr.GlobalVar` for its implementation
+and documentation.
+
+Local Variable
+~~~~~~~~~~~~~~~~~
+
+Local identifiers are prefixed by the :code:`%` sigil,
+such as ":code:`%local`". A local identifier always references
+a function argument or a variable bound in a `let` expression.
+A local variable will be scoped to the function where it
+appears or the `let` expression where it is bound, respectively.
+
+Suppose the local variable :code:`%a` has been defined in a scope
+and within that scope, a `let` expression binding to a variable
+:code:`%a` appears. This is permitted, as in most functional languages.
+In the scope of the `let` expression (the inner scope),
+the name :code:`%a` is "shadowed," meaning all references to
+:code:`%a` in the inner scope refer to the later defintion, while
+references to :code:`%a` in the outer scope continue to refer to
+the previous one.
+
+(Note that in Relay's implemetnation, each definition of a local variable
+creates a new :py:class:`~tvm.relay.expr.Var`, so a shadowed local variable,
+despite having the same name as one in an outer scope, will be a different
+object. This allows for comparing local variables by pointer identity with the
+knowledge that the same local variable object corresponds to a different binding site.)
+
+See :py:class:`~tvm.relay.expr.Var` for its implementation
+and documentation.
+
+Global Functions
+================
+
+A function is no different from a procedure or function in a
+typical programming language and generalizes the concept of a
+named subgraph.
+A function definition consists of a name, arguments, return type,
+type parameters, and any applicable type relations.
+A function's return type and the types of parameters may be omitted;
+Relay will attempt to infer the most general types where types
+are omitted.
+
+Functions defined in the manner described in this subsection are
+of global scope; anonymous functions will be discussed later, though
+their mechanics are nearly identical. Note that global functions may be
```

Comment:
Shadowing is tricky here, they won't actually be shadowed unless you are using the parser, and we should check with @joshpoll about that.

Reply:
Global functions can't be defined with a let expression, so they can't be shadowed this way.

Reply:
Oh I missed this point, I should remove the remark about shadowing the identifier